fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/mixin/event/lifecycle/ChunkHolderMixin.java

```
@@ -0,0 +1,140 @@
+	@Shadow
+	@Final
+	private HeightLimitView world;
+
+	@Shadow
+	private int lastTickLevel;
+
+	@Shadow
+	private volatile CompletableFuture<OptionalChunk<WorldChunk>> tickingFuture;
+
+	@Shadow
+	protected abstract void combineSavingFuture(CompletableFuture<?> savingFuture);
+
+	@Unique
+	private static final ChunkLevelType[] CHUNK_LEVEL_TYPES = ChunkLevelType.values(); // values() clones the internal array each call, so cache the return
+
+	@Unique
+	private CompletableFuture<Void> fullToBlockTickingEvent;
+
+	@Unique
+	private ChunkLevelType currentEventLevelType = INACCESSIBLE;
+
+	private ChunkHolderMixin(ChunkPos pos) {
+		super(pos);
+	}
+
+	/**
+	 * Handles INACCESSIBLE -> FULL for chunks that are immediately loaded and available. {@link ChunkGeneratingMixin} handles the rest.
+	 */
+	@Inject(method = "updateFutures", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/world/ChunkHolder;combineSavingFuture(Ljava/util/concurrent/CompletableFuture;)V", shift = At.Shift.AFTER, ordinal = 0))
+	private void updateFutures$inaccessibleToFull(ServerChunkLoadingManager chunkLoadingManager, Executor executor, CallbackInfo ci) {
+		if (this.getUncheckedOrNull(ChunkStatus.FULL) instanceof WorldChunk && this.currentEventLevelType == INACCESSIBLE) { // prevent duplicate events with ChunkGeneratingMixin
+			ServerChunkEvents.CHUNK_LEVEL_TYPE_CHANGE.invoker().onChunkLevelTypeChange((ServerWorld) world, this.pos, INACCESSIBLE, FULL);
+			this.currentEventLevelType = FULL;
+		}
+	}
+
+	/**
+	 * Handles FULL -> BLOCK_TICKING.
+	 */
+	@Inject(method = "updateFutures", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/world/ChunkHolder;combineSavingFuture(Ljava/util/concurrent/CompletableFuture;)V", shift = At.Shift.AFTER, ordinal = 1))
+	private void updateFutures$fullToBlockTicking(ServerChunkLoadingManager chunkLoadingManager, Executor executor, CallbackInfo ci) {
+		this.fullToBlockTickingEvent = this.tickingFuture.thenAccept(optionalChunk -> {
```

Exceptions thrown from listeners will be swallowed if you don't ever handle them explicitly.

---

I'm not sure what you mean by handle them explicitly.

---

Either with a try-catch or a `whenComplete` after this.